# **Laboratory Task 3**

Forward and Backward Propagation (ReLU)

Instruction: Perform a forward and backward propagation in python using the inputs from Laboratory Task 2

`x = np.array([1, 0, 1])
y = np.array([1])`

use relu as the activation function.

`# learning rate
lr = 0.001`

In [1]:
import numpy as np

# Inputs and expected output
x = np.array([1, 0, 1])
y = np.array([1])

In [2]:
# Learning rate
lr = 0.001

In [3]:
# Initialize weights and bias
np.random.seed(42)
W = np.random.randn(3, 1)   # 3 inputs → 1 output
b = np.random.randn(1)

In [4]:
# ReLU activation
def relu(z):
    return np.maximum(0, z)

def relu_derivative(z):
    return (z > 0).astype(float)

# Forward propagation
z = np.dot(x, W) + b
a = relu(z)

# Loss (Mean Squared Error)
loss = np.square(y - a).mean()

print("Forward Propagation")
print("Input:", x)
print("Weights:", W.flatten())
print("Bias:", b)
print("Z:", z)
print("Activated Output (a):", a)
print("Loss:", loss)

Forward Propagation
Input: [1 0 1]
Weights: [ 0.49671415 -0.1382643   0.64768854]
Bias: [1.52302986]
Z: [2.66743255]
Activated Output (a): [2.66743255]
Loss: 2.780331300528872


In [5]:
# Backward propagation
dz = (a - y) * relu_derivative(z)  # derivative of loss w.r.t z
dW = x.reshape(-1, 1) * dz
db = dz

# Update parameters
W -= lr * dW
b -= lr * db

print("\nBackward Propagation")
print("dW:", dW.flatten())
print("db:", db)
print("Updated Weights:", W.flatten())
print("Updated Bias:", b)


Backward Propagation
dW: [1.66743255 0.         1.66743255]
db: [1.66743255]
Updated Weights: [ 0.49504672 -0.1382643   0.64602111]
Updated Bias: [1.52136242]


**Observations:**

* The predicted output (2.67) is higher than the target (1), showing error exists.
* Gradient values are non-zero only for active inputs (1st and 3rd features).
* Both weights and bias are slightly reduced, nudging the output closer to the target.
* Learning rate (0.001) leads to very small adjustments → training is stable but gradual.
* Loss remains relatively high, indicating more updates are needed for convergence.